### ISSUE TITLE
[question] about the PrimaryKeyDiffScanner

### ISSUE THREAD

**Issue by aptly-io** [RxSwiftCommunity/RxGRDB#26]
I'm trying to represent an ordered list of persons in a UITableView section based on rxGRDB's events. The integer order table field defines a sorting order, the table uses soft-delete to handle changes. Following the example in rxGRDB's readme, the PrimaryKeyDiffScanner informs about changes in the persons table. 

* Does its implementation limit order by id only? How feasible is it to make a `SimpleDiffScanner`? Would the differ in GRDB (used to represent UITableView as well) be a good start for diffing collections?

* I don't understand the result for the following code snippet; I would expect 0 inserts and deletes, while 4 inserts and deletes are printed (4 being the db table actual row count).

```swift
     // ideally I want to use this request to get an ordered list of non-delete
     // let personsRequest = Person.order(Column("order")).filter(Column("deleted") == nil)

     // but first try to get a simple to work
     let personsRequest = Person.order(Column("id"))
     var initialPersons: [Person]?

     do {
         initialPersons = try DelayDb.queue.inDatabase { database in
             try personsRequest.fetchAll(database)
         }

         let scanner = try DelayDb.queue.inDatabase { database in
             try PrimaryKeyDiffScanner(
                 // **** I would expect inserting the initial table would result in no insertions/deletions later ****
                 database: database, request: personsRequest, initialRecords: initialPersons!)
         }

         let rowRequest = personsRequest.asRequest(of: Row.self)

         // The scanner is designed to feed the built-in RxSwift `scan` operator:
         rowRequest.rx
         .fetchAll(in: DelayDb.queue)
         .scan(scanner) { (scanner, rows) in
 //                    print(rows.map { $0.description } )     // I'm indeed seeing the right 'Row' values printed ...
             return scanner.diffed(from: rows)
         }
         .subscribe(onNext: { scanner in
             let diff = scanner.diff
             print("inserted \(diff.inserted.count) records")    // expect 0 iso 4
             print("updated \(diff.updated.count) records")
             print("deleted \(diff.deleted.count) records")   // expect 0 iso 4
         })
```
Output:
```
inserted 4 records
updated 0 records
deleted 4 records
```

Thanks for any clarifications on what I do wrong ... (using the cocoa pod version of rxGRDB)

**Comment by groue** [comment]
Hello @aptly-io,

> I'm trying to represent an ordered list of persons in a UITableView section based on rxGRDB's events.

All persons are gathered in a single section, very good.

> The integer order table field defines a sorting order, the table uses soft-delete to handle changes.

By "soft delete", you mean that you set a "deleted" flag instead of actually deleting persons. Fine.

> Following the example in rxGRDB's readme, the PrimaryKeyDiffScanner informs about changes in the persons table.

Yes. But it also [says](https://github.com/RxSwiftCommunity/RxGRDB/blob/master/README.md#primarykeydiffscanner):

> PrimaryKeyDiffScanner computes diffs between collections **whose order does not matter**.

Because PrimaryKeyDiffScanner does not care about ordering, I don't think it is a good fit for UITableView.

RxGRDB ships with a [demo app](https://github.com/RxSwiftCommunity/RxGRDB/tree/master/Documentation/RxGRDBDemo) that use RxGRDB to feed a table view. It does not use PrimaryKeyDiffScanner, and instead uses another diffing algorithm, that is suited for table views.

You should have a look at it.

> Does its implementation limit order by id only? How feasible is it to make a SimpleDiffScanner? Would the differ in GRDB (used to represent UITableView as well) be a good start for diffing collections?

Some answer: PrimaryKeyDiffScanner is not the correct tool for this job, and you won't be able to derive any "SimpleDiffScanner" from it.

There is no such thing as a "simple diff". You need a *specific* diffing algorithm suited for table views, such as the one used by the [demo app](https://github.com/RxSwiftCommunity/RxGRDB/tree/master/Documentation/RxGRDBDemo).

> I don't understand the result for the following code snippet; I would expect 0 inserts and deletes, while 4 inserts and deletes are printed (4 being the db table actual row count).

Even if I think that you will not use PrimaryKeyDiffScanner in the end, there's no reason for not explaining the behavior you see.

And... this looks odd to me, too. I'll have a deeper look. Meanwhile, would you please run the demo app, tell me if it looks that if fits your needs, and if you are able to adapt it for your own app?

**Comment by groue** [comment]
I should add that RxGRDB does not yet talk to [RxSwiftCommunity/RxDataSources ](https://github.com/RxSwiftCommunity/RxDataSources). When it does, I expect that it will become easier to deal with table views.

**Comment by groue** [comment]
Another piece of information that may interest you: GRDB itself ships with [FetchedRecordsController](https://github.com/groue/GRDB.swift/blob/master/README.md#fetchedrecordscontroller), which is not RX at all, but is 100% meant for automatic updates of a table view as the results of a request change. And it's very easy to setup.

Database observation is a feature of GRDB. RxGRDB only adds a little RX flavor. There are cool stuff in the mothership library as well.

**Comment by aptly-io** [comment]
@groue Thank you for the clarifications. Yes, the soft-delete uses the `deleted` field to mark rows that are no longer to be used. The id is only for referencing a `Person`, I don't intend to use it for sorting; I plan the `order` field for that.

I will look into the demo app for more inspiration. It is not trivial to get that to work (as a relatively newcomer to Xcode and its building logic).

I had a short look into RxDataSources but these don't seem to be connected to a DB and assume one is having direct access to ones own UITableView? I'm trying to use (Eureka)[https://github.com/xmartlabs/Eureka] `MultivaluedSection`s however (and yes a relatively hard time to fit it with RGDB). The future plan was to benefit from rxRGDB's notifications on db updates (e.g. a different component using CloudKit would commit changes to the db, then rxGRDB would trigger table view(s) to update automagically).

**Comment by aptly-io** [comment]
@groue Yes, I played with the `RGDB.FetchedRecordsController` when I managed my own `UITableView`. As written above (we cross posted), I try to benefit from the `Eureka` library however. 
I feel like the data source controllers want to much control over the `UITableView` to make these libraries (RGDB and Eureka) play well together ...

**Comment by groue** [comment]
Hmmm. Yes, it's a not trivial topic at all. I'm curious to know what your stack eventually looks like. And if you have any more question, or suggestions, be welcome!

**Comment by aptly-io** [comment]
@groue Thanks for all the advice. 
Very initial results from converting the `RxGRDB.PrimaryKeyDiffScanner` into an `ExtendedDiffScanner` using the Differ framework like in the RxGRDBDemo's `PlayersViewController` seem to return the expected results. (much investigation for my correct understanding is still ahead).
